### ISSUE TITLE
DIscussion: remove the "write backend"?

### ISSUE THREAD

**Issue by mmd-osm** [zerebubuth/openstreetmap-cgimap#197]
Currently, cgimap has two completely separate implementations for the APIDB backend: one targeting a read-only mirror, and another one a writable database (which allows us to use temporary tables to store intermediate object ids and object id/version numbers respectively). 

At this time, osm.org uses the "read-only" backend in production. Recent attempts to switch to the "write backend" exhibited two previously unknown issues, which have since been fixed.

Presumably, the write backend was written with the intent to reduce cgimap<-->db traffic, potentially having a better response time. I was curious to see, if this results in significant speed ups.

Unfortunately, the results were quite disappointing: for larger bboxes with lots of ways and relations (280 relations, 5500 ways, 49000 nodes), both implementations took around 1s with a difference less than 100ms. In some cases, the write backend was even slightly slower (also in a 100ms range). Someone on Slack US even [reported a massive performance impact](https://osmus.slack.com/archives/C029HV951/p1562577921147800?thread_ts=1562051736.054200&cid=C029HV951) when using the write backend.

This is a bit of an uncomfortable situation: we have two implementations for the same functionality with rather insignificant performance differences. On the other hand, having two implementations adds long term maintenance costs to keep both in sync, etc. Also, it makes it harder for new people to work on the code base.

Long story short: my radical suggestion would be to remove the write backend, and focus on the read-only backend only.

This change would remove about 1000 lines of code: https://github.com/mmd-osm/openstreetmap-cgimap/commit/b323ee7454618ab5a34a3c8346dd0ee894d47ba6

**Comment by tomhughes** [comment]
The write backend was actually the original implementation - the readonly one was added later when we needed to be able to run against a mirror during a move of the primary server if I recall correctly.

**Comment by mmd-osm** [comment]
Ok, so the original motivation was not performance related. When did we actually switch to the readonly implementation in production?

**Comment by tomhughes** [comment]
Quite the opposite - the readonly backend was expected to be slower. It was a compromise approach to get something that could work when temporary tables weren't available.

I'm not sure about the exact history. I think readonly was originally used while the main server was being moved from London to York and after that we certainly intended to switch back - no idea if we did but certainly at some point we accidentally wound up running in readonly mode without intending to.

**Comment by mmd-osm** [comment]
Just adding a few measurements for the `/map` call I mentioned earlier on.

The read only backend needs 8 calls, while the read write backend needs 17.

It seems that combination of storing results in a temporary table and using that the same temp table in a LEFT OUTER join is just too costly, compared to simply passing a list of object ids back and forth between the db and cgimap.

### Read only

```
openstreetmap=# SELECT query, calls, total_time, rows, 100.0 * shared_blks_hit /
               nullif(shared_blks_hit + shared_blks_read, 0) AS hit_percent
          FROM pg_stat_statements ORDER BY total_time DESC;
                                                                       query                                                                        | calls | total_time | rows  |     hit_percent      
----------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------+-------+----------------------
 SELECT n.id, n.latitude, n.longitude, n.visible, to_char(n.timestamp,'YYYY-MM-DD"T"HH24:MI:SS"Z"') AS timestamp, n.changeset_id, n.version, array_.|     1 |    241.342 | 48978 | 100.0000000000000000
.agg(t.k) as tag_k, array_agg(t.v) as tag_v FROM current_nodes n LEFT JOIN current_node_tags t ON n.id=t.node_id WHERE n.id = ANY($1) GROUP BY n.id.|       |            |       | 
. ORDER BY n.id                                                                                                                                     |       |            |       | 
 SELECT w.id, w.visible, to_char(w.timestamp,'YYYY-MM-DD"T"HH24:MI:SS"Z"') AS timestamp, w.changeset_id, w.version, t.keys as tag_k, t.values as ta.|     1 |     67.324 |  5486 | 100.0000000000000000
.g_v, wn.node_ids as node_ids FROM current_ways w LEFT JOIN LATERAL (SELECT array_agg(k) as keys, array_agg(v) as values FROM current_way_tags WHER.|       |            |       | 
.E w.id=way_id) t ON true LEFT JOIN LATERAL (SELECT array_agg(node_id) as node_ids FROM (SELECT node_id FROM current_way_nodes WHERE w.id=way_id OR.|       |            |       | 
.DER BY sequence_id) x) wn ON true WHERE w.id = ANY($1) ORDER BY w.id                                                                               |       |            |       | 
 SELECT DISTINCT wn.way_id AS id FROM current_way_nodes wn WHERE wn.node_id = ANY($1)                                                               |     1 |     62.319 |  5486 | 100.0000000000000000
 SELECT DISTINCT rm.relation_id AS id FROM current_relation_members rm WHERE rm.member_type = 'Way' AND rm.member_id = ANY($1)                      |     3 |     50.229 |   280 | 100.0000000000000000
 SELECT DISTINCT wn.node_id AS id FROM current_way_nodes wn WHERE wn.way_id = ANY($1)                                                               |     1 |     40.649 | 48977 | 100.0000000000000000
 SELECT id FROM current_nodes WHERE tile = ANY($1) AND latitude BETWEEN $2 AND $3 AND longitude BETWEEN $4 AND $5 AND visible = true LIMIT $6       |     1 |     34.332 | 48556 | 100.0000000000000000
 SELECT r.id, r.visible, to_char(r.timestamp,'YYYY-MM-DD"T"HH24:MI:SS"Z"') AS timestamp, r.changeset_id, r.version, t.keys as tag_k, t.values as ta.|     1 |      3.413 |   280 | 100.0000000000000000
.g_v, rm.types as member_types, rm.ids as member_ids, rm.roles as member_roles FROM current_relations r LEFT JOIN LATERAL (SELECT array_agg(k) as k.|       |            |       | 
.eys, array_agg(v) as values FROM current_relation_tags WHERE r.id=relation_id) t ON true LEFT JOIN LATERAL (SELECT array_agg(member_type) as types.|       |            |       | 
., array_agg(member_role) as roles, array_agg(member_id) as ids FROM (SELECT * FROM current_relation_members WHERE r.id=relation_id ORDER BY sequen.|       |            |       | 
.ce_id) x) rm ON true WHERE r.id = ANY($1) ORDER BY r.id                                                                                            |       |            |       | 
 SELECT c.id, to_char(c.created_at,'YYYY-MM-DD"T"HH24:MI:SS"Z"') AS created_at, to_char(c.closed_at, 'YYYY-MM-DD"T"HH24:MI:SS"Z"') AS closed_at, c..|     1 |      0.001 |     0 |                     
.min_lat, c.max_lat, c.min_lon, c.max_lon, c.num_changes, t.keys as tag_k, t.values as tag_v, cc.author_id as comment_author_id, cc.display_name as.|       |            |       | 
. comment_display_name, cc.body as comment_body, cc.created_at as comment_created_at FROM changesets c LEFT JOIN LATERAL (SELECT array_agg(k) AS ke.|       |            |       | 
.ys, array_agg(v) AS values FROM changeset_tags WHERE c.id=changeset_id ) t ON true LEFT JOIN LATERAL (SELECT array_agg(author_id) as author_id, ar.|       |            |       | 
.ray_agg(display_name) as display_name, array_agg(body) as body, array_agg(created_at) as created_at FROM (SELECT cc.author_id, u.display_name, cc..|       |            |       | 
.body, to_char(cc.created_at,'YYYY-MM-DD"T"HH24:MI:SS"Z"') AS created_at FROM changeset_comments cc JOIN users u ON cc.author_id = u.id where cc.ch.|       |            |       | 
.angeset_id=c.id AND cc.visible ORDER BY cc.created_at) x )cc ON true WHERE c.id = ANY($1)                                                          |       |            |       | 


```
### Read-write

```

openstreetmap=# SELECT query, calls, total_time, rows, 100.0 * shared_blks_hit /
               nullif(shared_blks_hit + shared_blks_read, 0) AS hit_percent
          FROM pg_stat_statements ORDER BY total_time DESC;
                                                                       query                                                                        | calls | total_time | rows  |     hit_percent      
----------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------+-------+----------------------
 SELECT n.id, n.latitude, n.longitude, n.visible, to_char(n.timestamp,?) AS timestamp, n.changeset_id, n.version, array_agg(t.k) as tag_k, array_ag.|     1 |    259.185 | 48978 | 100.0000000000000000
.g(t.v) as tag_v FROM current_nodes n JOIN tmp_nodes tn ON n.id=tn.id LEFT JOIN current_node_tags t ON n.id=t.node_id GROUP BY n.id                 |       |            |       | 
 INSERT INTO tmp_ways SELECT DISTINCT wn.way_id AS id FROM current_way_nodes wn JOIN tmp_nodes tn ON wn.node_id = tn.id LEFT JOIN tmp_ways tw ON wn.|     1 |    102.558 |  5486 | 100.0000000000000000
..way_id = tw.id WHERE tw.id IS NULL                                                                                                                |       |            |       | 
 INSERT INTO tmp_nodes SELECT DISTINCT wn.node_id AS id FROM tmp_ways tw JOIN current_way_nodes wn ON tw.id = wn.way_id LEFT JOIN tmp_nodes tn ON w.|     1 |     79.707 |   422 | 100.0000000000000000
.n.node_id = tn.id WHERE tn.id IS NULL                                                                                                              |       |            |       | 
 SELECT w.id, w.visible, w.version, w.changeset_id, to_char(w.timestamp,?) AS timestamp, t.keys as tag_k, t.values as tag_v, wn.node_ids as node_id.|     1 |     78.196 |  5486 | 100.0000000000000000
.s FROM current_ways w JOIN tmp_ways tw ON w.id=tw.id LEFT JOIN LATERAL (SELECT array_agg(k) AS keys, array_agg(v) AS values FROM current_way_tags .|       |            |       | 
.WHERE w.id=way_id ) t ON ? LEFT JOIN LATERAL (SELECT array_agg(node_id) AS node_ids from (SELECT * FROM current_way_nodes WHERE w.id=way_id ORDER .|       |            |       | 
.BY sequence_id) x ) wn ON ?                                                                                                                        |       |            |       | 
 INSERT INTO tmp_nodes SELECT id FROM current_nodes WHERE tile = ANY($1) AND latitude BETWEEN $2 AND $3 AND longitude BETWEEN $4 AND $5 AND visible.|     1 |      59.14 | 48556 | 100.0000000000000000
. = ? LIMIT $6                                                                                                                                      |       |            |       | 
 INSERT INTO tmp_relations SELECT DISTINCT rm.relation_id FROM tmp_nodes tn JOIN current_relation_members rm ON (tn.id = rm.member_id AND rm.member.|     1 |     38.387 |     0 | 100.0000000000000000
._type=?) LEFT JOIN tmp_relations tr ON rm.relation_id = tr.id WHERE tr.id IS NULL                                                                  |       |            |       | 
 INSERT INTO tmp_relations SELECT DISTINCT rm.relation_id AS id FROM tmp_ways tw JOIN current_relation_members rm ON (tw.id = rm.member_id AND rm.m.|     1 |     17.213 |   280 | 100.0000000000000000
.ember_type=?) LEFT JOIN tmp_relations tr ON rm.relation_id = tr.id WHERE tr.id IS NULL                                                             |       |            |       | 
 CREATE TEMPORARY TABLE tmp_nodes (id bigint PRIMARY KEY)                                                                                           |     2 |      5.416 |     0 |  98.8479262672811060
 SELECT r.id, r.visible, r.version, r.changeset_id, to_char(r.timestamp,?) AS timestamp, t.keys as tag_k, t.values as tag_v, rm.types as member_typ.|     1 |      3.644 |   280 | 100.0000000000000000
.es, rm.ids as member_ids, rm.roles as member_roles FROM current_relations r JOIN tmp_relations tr ON tr.id=r.id LEFT JOIN LATERAL (SELECT array_ag.|       |            |       | 
.g(k) AS keys, array_agg(v) AS values FROM current_relation_tags WHERE r.id=relation_id ) t ON ? LEFT JOIN LATERAL (SELECT array_agg(member_type) A.|       |            |       | 
.S types, array_agg(member_id) AS ids, array_agg(member_role) AS roles FROM ( SELECT * FROM current_relation_members WHERE r.id=relation_id ORDER B.|       |            |       | 
.Y sequence_id) x)rm ON ?                                                                                                                           |       |            |       | 
 CREATE TEMPORARY TABLE tmp_ways (id bigint PRIMARY KEY)                                                                                            |     2 |      1.711 |     0 | 100.0000000000000000
 CREATE TEMPORARY TABLE tmp_changesets (id bigint PRIMARY KEY)                                                                                      |     2 |      1.688 |     0 | 100.0000000000000000
 CREATE TEMPORARY TABLE tmp_relations (id bigint PRIMARY KEY)                                                                                       |     2 |      1.673 |     0 | 100.0000000000000000
 CREATE TEMPORARY TABLE tmp_historic_nodes (node_id bigint, version bigint, PRIMARY KEY (node_id, version))                                         |     2 |      1.556 |     0 | 100.0000000000000000
 CREATE TEMPORARY TABLE tmp_historic_ways (way_id bigint, version bigint, PRIMARY KEY (way_id, version))                                            |     2 |      1.497 |     0 | 100.0000000000000000
 CREATE TEMPORARY TABLE tmp_historic_relations (relation_id bigint, version bigint, PRIMARY KEY (relation_id, version))                             |     2 |      1.159 |     0 | 100.0000000000000000
 INSERT INTO tmp_relations SELECT DISTINCT rm.relation_id FROM tmp_relations tr JOIN current_relation_members rm ON (tr.id = rm.member_id AND rm.me.|     1 |      0.005 |     0 | 100.0000000000000000
.mber_type=?) LEFT JOIN tmp_relations xr ON rm.relation_id = xr.id WHERE xr.id IS NULL                                                              |       |            |       | 
 SELECT c.id, to_char(c.created_at,?) AS created_at, to_char(c.closed_at, ?) AS closed_at, c.min_lat, c.max_lat, c.min_lon, c.max_lon, c.num_change.|     1 |          0 |     0 |                     
.s, t.keys as tag_k, t.values as tag_v, cc.author_id as comment_author_id, cc.display_name as comment_display_name, cc.body as comment_body, cc.cre.|       |            |       | 
.ated_at as comment_created_at FROM changesets c JOIN tmp_changesets tc ON tc.id=c.id LEFT JOIN LATERAL (SELECT array_agg(k) AS keys, array_agg(v) .|       |            |       | 
.AS values FROM changeset_tags WHERE c.id=changeset_id ) t ON ? LEFT JOIN LATERAL (SELECT array_agg(author_id) as author_id, array_agg(display_name.|       |            |       | 
.) as display_name, array_agg(body) as body, array_agg(created_at) as created_at FROM (SELECT cc.author_id, u.display_name, cc.body, to_char(cc.cre.|       |            |       | 
.ated_at,?) AS created_at FROM changeset_comments cc JOIN users u ON cc.author_id = u.id where cc.changeset_id=c.id AND cc.visible ORDER BY cc.crea.|       |            |       | 
.ted_at) x )cc ON ?                                                                                                                                 |       |            |       | 




```

**Comment by tomhughes** [comment]
I'm not saying the write backend is faster, just that it was the expectation. I don't know if it was ever measured.

It's also true that there have been lots of optimisations of the queries since which may have changed things, especially if they only apply to the readonly backend.

**Comment by mmd-osm** [comment]
> It's also true that there have been lots of optimisations of the queries since which may have changed things, especially if they only apply to the readonly backend.

That's true. About two years ago, both backends have seen some optimization effort to cut down on the number of single select statements. I think, both backends are very similar on how they go about collecting the data, just that the one keeps the ids in a temporary table, and the other passes them back and forth.

**Comment by mmd-osm** [comment]
So I merged this for now. Hopefully, the reasoning why the code was removed is good enough for someone looking into this some time in the future.
